src/invidious/helpers/youtube_api.cr

```diff
@@ -2,120 +2,398 @@
-  # (this is required for channel info, playlist and community, e.g)
-  if params != ""
-    data["params"] = params
+    # :ditto:
+    def version : String
+      HARDCODED_CLIENTS[@client_type][:version]
+    end
+
+    # :ditto:
+    def api_key : String
+      HARDCODED_CLIENTS[@client_type][:api_key]
+    end
+
+    # Convert to string, for logging purposes
+    def to_s
+      return {
+        client_type:  self.name,
+        region:       @region,
+        proxy_region: @proxy_region,
+      }.to_s
+    end
   end
 
-  return _youtube_api_post_json("/youtubei/v1/browse", data)
-end
+  # Default client config, used if nothing is passed
+  DEFAULT_CLIENT_CONFIG = ClientConfig.new
 
-####################################################################
-# request_youtube_api_search(search_query, params, region)
-#
-# Requests the youtubei/v1/search endpoint with the required headers
-# and POST data in order to get a JSON reply. As the search results
-# vary depending on the region, a region code can be specified in
-# order to get non-US results.
-#
-# The requested data is a search string, with some additional
-# paramters, formatted as a base64 string.
-#
-def request_youtube_api_search(search_query : String, params : String, region = nil)
-  # JSON Request data, required by the API
-  data = {
-    "query"   => search_query,
-    "context" => make_youtube_api_context(region),
-    "params"  => params,
-  }
+  ####################################################################
+  # make_context(client_config)
+  #
+  # Return, as a Hash, the "context" data required to request the
+  # youtube API endpoints.
+  #
+  private def make_context(client_config : ClientConfig | Nil) : Hash
+    # Use the default client config if nil is passed
+    client_config ||= DEFAULT_CLIENT_CONFIG
 
-  return _youtube_api_post_json("/youtubei/v1/search", data)
-end
+    client_context = {
+      "client" => {
+        "hl"            => "en",
+        "gl"            => client_config.region || "US", # Can't be empty!
+        "clientName"    => client_config.name,
+        "clientVersion" => client_config.version,
+      },
+    }
 
-####################################################################
-# _youtube_api_post_json(endpoint, data)
-#
-# Internal function that does the actual request to youtube servers
-# and handles errors.
-#
-# The requested data is an endpoint (URL without the domain part)
-# and the data as a Hash object.
-#
-def _youtube_api_post_json(endpoint, data)
-  # Send the POST request and parse result
-  response = YT_POOL.client &.post(
-    "#{endpoint}?key=#{HARDCODED_API_KEY}",
-    headers: HTTP::Headers{"content-type" => "application/json; charset=UTF-8"},
-    body: data.to_json
+    # Used for age-restriction bypass
+    # Use "Case-when" to allow other type of client in the future
+    case client_config.client_type
+    when ClientType::WebEmbed
+      client_context["client"]["clientScreen"] = "EMBED"
+    end
```

Comment:
Using this for `WEB_EMBEDDED_PLAYER` is useless. tier 2 videos like `Tq92D6wQ1mg` still fail in this case. This option should be used with `WEB` instead. Also, note that setting this for all videos don't work since some videos like `tf2U5Vyj0oU` fail with this set

Reply:
The primary idea behind that is to mimick as much as possible the behavior of a "normal" youtube client, to (try to) reduce as much as possible the risk of throttling (or any other annoyance) from google.

Reply:
Still, this needs to used this with `WEB` to bypass age-gate. Also `client_context["thirdParty"]` needs to be set for robustness when using `EMBED`. See https://github.com/yt-dlp/yt-dlp/pull/575#issuecomment-887739287